trieste/acquisition/function.py

```diff
@@ -524,3 +527,194 @@ def prepare_acquisition_function(
+        """
+        single_builder = self
+
+        class _Anon(BatchAcquisitionFunctionBuilder):
+            def prepare_acquisition_function(
+                self, datasets: Mapping[str, Dataset], models: Mapping[str, ProbabilisticModel]
+            ) -> AcquisitionFunction:
+                return single_builder.prepare_acquisition_function(datasets[tag], models[tag])
+
+            def __repr__(self) -> str:
+                return f"{single_builder!r} using tag {tag!r}"
+
+        return _Anon()
+
+    @abstractmethod
+    def prepare_acquisition_function(
+        self, dataset: Dataset, model: ProbabilisticModel
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data to use to build the acquisition function.
+        :param model: The model over the specified ``dataset``.
+        :return: An acquisition function.
+        """
+
+
+def predict_independent_f_samples_with_reparametrisation_trick(
+        model: ProbabilisticModel, at: QueryPoints, eps: tf.Tensor
+) -> tf.Tensor:
+    """
+    Returns independent samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    N is the number of query points and L is the number of latent processes.
+    :param model:
+    :param at: N query points [N, D]
+    :param eps: a [S, L] tensor of N(0, 1) i.i.d. samples
+    :return: a [S, N, L] tensor
+    """
+    assert len(at.shape) == 2
+    mean, cov = model.predict(at)   # both [N, L]
+    return mean[None, ...] + tf.math.sqrt(cov)[None, ...] * eps[:, None, :]  # [S, N, L]
+
+
+def predict_batch_f_samples_with_reparametrisation_trick(
+        model: ProbabilisticModel, at: QueryPoints, eps: tf.Tensor
+) -> tf.Tensor:
+    """
+    Returns batch-correlated samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    the batch size B must be compatible between eps and at,
+    N is the number of batches of query points
+    and L is the number of latent processes.
+    :param model:
+    :param at: N batches of query points [N, B, D]
+    :param eps: a [S, B, L] tensor of N(0, 1) i.i.d. samples
+    :return: a [S, N, B, L] tensor
+    """
+
+    assert at.shape[1] == eps.shape[1]
+
+    mean, cov = model.predict(at, full_cov=True)
+    # mean: [N, B, L]
+    # cov: [N, L, B, B]
+    mean_for_sample = tf.linalg.adjoint(mean)  # [N, L, B]
+    mean_shape = tf.shape(mean_for_sample)
+    B = mean_shape[-1]
+
+    jittermat = (
+            tf.eye(B, batch_shape=mean_shape[:-1], dtype=default_float()) * default_jitter()
+    )  # [N, L, B, B]
+
+    eps = tf.transpose(eps, [2, 1, 0])  # [S, B, L] -> [L, B, S]
+
+    chol = tf.linalg.cholesky(cov + jittermat)  # [N, L, B, B]
+    samples = mean_for_sample[..., None] + tf.linalg.matmul(chol, eps)  # [N, L, B, S]
+    return tf.transpose(samples, [3, 0, 2, 1])
+
+
+class MonteCarloExpectedImprovement(SingleModelAcquisitionBuilder):
```

Comment:
can you clarify for me: is there actually such a thing as "Monte-Carlo Expected Improvement", or is there just "Expected Improvement" achieved via Monte Carlo sampling with the reparametrization trick? Cos if they're the same, why not merge them into a single acquisition and just use the analytic form where it's possible and the MC trick otherwise (which if i understand correctly is the batch case)? Well, I guess to keep batch and non-batch separate, we'll have to keep them separate, but in that case is it more
"ExpectedImprovement" vs "BatchExpectedImprovement"?

Reply:
ah, ok, I've scrolled down a bit to "BatchMonteCarloExpectedImprovement", which partially answers my question. So my next question is then: what does "MonteCarloExpectedImprovement" that "ExpectedImprovement" doesn't?

and the same qu for `MonteCarloNegativeLowerConfidenceBound` vs `NegativeLowerConfidenceBound`